### ISSUE TITLE
Hidden Amber Wolverine - Founders could still obtain vesting NFTs even when they have been intentionally removed by the contract owner

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-09-nounsbuilder-judging#259]
Hidden Amber Wolverine

high

# Founders could still obtain vesting NFTs even when they have been intentionally removed by the contract owner

## Summary
The incorrect implementation of ````updateFounders()```` would cause that founders could still claim vesting NFTs even when they have been intentionally removed by the contract owner.

## Vulnerability Detail
The issue arises on L412, the ````baseTokenId```` should begin with ````reservedUntilTokenId % 100````.
```diff
File: src\token\Token.sol
375:     function updateFounders(IManager.FounderParams[] calldata newFounders) external onlyOwner {
...
393:         unchecked {
394:             // for each existing founder:
395:             for (uint256 i; i < cachedFounders.length; ++i) {
...
407: 
408:                 // using the ownership percentage, get reserved token percentages
409:                 uint256 schedule = 100 / cachedFounder.ownershipPct;
410: 
411:                 // Used to reverse engineer the indices the founder has reserved tokens in.
-412:                 uint256 baseTokenId; // @audit should begin with reservedUntilTokenId
+412:                 uint256 baseTokenId = reservedUntilTokenId % 100;
413: 
414:                 for (uint256 j; j < cachedFounder.ownershipPct; ++j) {
415:                     // Get the next index that hasn't already been cleared
416:                     while (clearedTokenIds[baseTokenId] != false) {
417:                         baseTokenId = (++baseTokenId) % 100;
418:                     }
419: 
420:                     delete tokenRecipient[baseTokenId];
421:                     clearedTokenIds[baseTokenId] = true;
422: 
423:                     emit MintUnscheduled(baseTokenId, i, cachedFounder);
424: 
425:                     // Update the base token id
426:                     baseTokenId = (baseTokenId + schedule) % 100;
427:                 }
428:             }
429:         }
...
437:     }
```
 let's say the ````reservedUntilTokenId == 10```` and ````founderPct = 5````, when ````_addFounders()```` in ````initialize()````, the following ````tokenRecipient````s would be set.
```solidity
tokenRecipient[10]
tokenRecipient[30]
tokenRecipient[50]
tokenRecipient[70]
tokenRecipient[90]
```
But in ````updateFounders()````, as ````baseTokenId```` begin with ````0````, the actual cleared ````tokenRecipient````s are
```solidity
tokenRecipient[0]
tokenRecipient[20]
tokenRecipient[40]
tokenRecipient[60]
tokenRecipient[80]
```
We can see, none of previous valid ````tokenRecipient```` info is cleared. Hence, founders could still obtain those vesting NFTs.

## Impact
Founders could still claim vesting NFTs even when they have been intentionally removed.

## Code Snippet
https://github.com/sherlock-audit/2023-09-nounsbuilder/blob/main/nouns-protocol/src/token/Token.sol#L412

## Tool used

Manual Review

## Recommendation
see Vulnerability Detail

**Comment by huye0728** [comment]
escalate

This is not dup of https://github.com/sherlock-audit/2023-09-nounsbuilder-judging/issues/42, with different affected function, different impact, different fix.

**Comment by ydspa** [comment]
Escalate

This is not dup of https://github.com/sherlock-audit/2023-09-nounsbuilder-judging/issues/42, with different affected function, different impact, different fix.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> This is not dup of https://github.com/sherlock-audit/2023-09-nounsbuilder-judging/issues/42, with different affected function, different impact, different fix.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by nevillehuang** [comment]
Will review all duplicates under #42, so please check deduplication comments [here](https://discord.com/channels/812037309376495636/1176899915411099719/1184507360765546556) in the mean time.

**Comment by nevillehuang** [comment]
See comment [here](https://github.com/sherlock-audit/2023-09-nounsbuilder-judging/issues/42#issuecomment-1857558554)

**Comment by Czar102** [comment]
I agree with the Lead Judge, planning to leave the issue as is. @ydspa please let me know if the logic in the above comment makes sense.

**Comment by Czar102** [comment]
Result:
High
Duplicate of #42

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [ydspa](https://github.com/sherlock-audit/2023-09-nounsbuilder-judging/issues/259/#issuecomment-1853926567): rejected
